macros/core/externalData.pl

```diff
@@ -0,0 +1,123 @@
+################################################################################
+# WeBWorK Online Homework Delivery System
+# Copyright &copy; 2000-2022 The WeBWorK Project, https://github.com/openwebwork
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of either: (a) the GNU General Public License as published by the
+# Free Software Foundation; either version 2, or (at your option) any later
+# version, or (b) the "Artistic License" which comes with this package.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See either the GNU General Public License or the
+# Artistic License for more details.
+################################################################################
+
+=head1 NAME
+
+externalData.pl - Macro that reads and writes data from a problem to the database.
+This is useful for using data common to multiple problems.
+
+=cut
+
+sub _init_externalData { }
+
+=head2 store_number_list
+
+This is a answer checker that checks only that there are a list of real number values
+and the purpose is to store in the database.
+
+=head3 Example
+
+Make a List object, but the contents doesn't matter.
+
+  $ans = List(0);
+
+Then in the visible problem block add the following
+
+  BEGIN_PGML
+  Enter a list of values as a vector.
+
+  [_____]{$ans->cmp(list_checker => store_number_list(), key => 'my_list')}
+  END_PGML
+
+Note that you will need to add a key (this should be unique for the problem set)
+
+=cut
+
+sub store_number_list {
+	return sub {
+		my ($correct, $student, $ansHash, $value) = @_;
+		my @errors;    # stores error messages
+
+		# Make sure that the key is set
+		my $key = $ansHash->{key};
+		warn 'You must pass in the key to save the data.' unless $key;
+
+		# Get the answer name for this input and store the name and key.
+		RECORD_EXTRA_ANSWERS("_ext_data:$ansHash->{ans_label}:numeric_list:$key");
+
+		# check that all numbers are real.
+		my $all_real = 1;
+		$all_real = $all_real && Value::isRealNumber($_) for (@$student);
+		push(@errors, 'One of the numbers is not a real number') unless $all_real;
+
+		return ($all_real ? scalar(@$student) : 0, @errors);
```

Comment:
It seems that this would be a lot cleaner and readable if it were:
```suggestion
		for (@$student) {
			return 0, 'One of the numbers is not a real number' unless Value::isRealNumber($_);
		}

		return scalar(@$student), @errors;
```

Although I don't think you are using the correct check with `Value::isRealNumber($_)`.  That is unless you intend for formulas involving variables to be allowed as entries in the list.  With that check an answer of `1,2,x,x^2+4` will be counted as correct.  If this is intended then you will need to change https://github.com/openwebwork/webwork2/pull/1879 to accept that sort of answer.  An answer like the example given causes the warning `Argument " x" isn't numeric in addition (+) at /opt/webwork/webwork2/lib/WeBWorK/Utils/ProblemProcessing.pm line 485.`   I doubt you did intend to allow that.  So the check should be `Value::classMatch($_, 'Real')` instead.

Reply:
This is intended for Reals not formulas with x.